### PR TITLE
Refac loss compute

### PR DESCRIPTION
**Pull request by vince62s**
Get rid of the sharding of batches (to minimize the memory usage during loss compute) since one can use the "tokens" batch option and adjust the size according to the GPU memory.

Refactor the Loss Compute to simplify the different cases (NMT, LM, copy_attn or not, ...)

Align the constructor on the other "-from_opts()' methods.

### CODE REVIEW

File: onmt/utils/loss.py [hunk 1]
```diff
@@ -110,51 +29,140 @@ class LossComputeBase(nn.Module):
...
+        self.tgt_shift_index = tgt_shift_index
+        self.copy_attn = copy_attn
+        self.vocab = vocab  # target vocab for copy_attn need
+
+    @classmethod
+    def from_opts(cls, opt, model, vocab, train=True):
+        """
+        Returns a subclass which wraps around an nn.Module subclass
+        (such as nn.NLLLoss) which defines the loss criterion. The LossCompute
+        object allows this loss to be computed in shards and passes relevant
```
Comment (francoishernandez) [inline onmt/utils/loss.py:50]: no more sharding?

File: onmt/utils/loss.py [hunk 2]
```diff
@@ -110,51 +29,140 @@ class LossComputeBase(nn.Module):
...
+        data to a Statistics object which handles training/validation logging.
+        Currently, the NMTLossCompute class handles all loss computation except
+        for when using a copy mechanism.
```
Comment (francoishernandez) [inline onmt/utils/loss.py:53]: Needs updating as well.

File: onmt/utils/loss.py [hunk 3]
```diff
@@ -110,51 +29,140 @@ class LossComputeBase(nn.Module):
...
+                    opt.label_smoothing, len(vocab),
+                    ignore_index=padding_idx
+                )
+            elif isinstance(model.generator[-1], LogSparsemax):
+                criterion = SparsemaxLoss(ignore_index=padding_idx,
+                                          reduction='sum')
+            else:
+                criterion = nn.NLLLoss(ignore_index=padding_idx,
+                                       reduction='sum')
+            vocab = vocab
```
Comment (francoishernandez) [inline onmt/utils/loss.py:86]: ?

File: onmt/utils/loss.py [hunk 4]
```diff
@@ -110,51 +29,140 @@ class LossComputeBase(nn.Module):
...
+                "order to use --lambda_coverage != 0"
+
+        tgt_shift_idx = 1 if opt.model_task == ModelTask.SEQ2SEQ else 0
+
+        if opt.copy_attn:
+            criterion = onmt.modules.CopyGeneratorLoss(
+                len(vocab), opt.copy_attn_force,
+                unk_index=unk_idx, ignore_index=padding_idx
+            )
+            vocab = vocab
```
Comment (francoishernandez) [inline onmt/utils/loss.py:72]: ?

File: onmt/utils/loss.py [hunk 5]
```diff
@@ -184,22 +191,59 @@ def __call__(self,
...
+        elif self.normalization == "sents":
+            normfactor = batch['tgt'].size(1)
+
+        return loss / float(normfactor), stats
 
     def _stats(self, loss, scores, target):
         """
         Args:
-            loss (:obj:`FloatTensor`): the loss computed by the loss criterion.
+            loss (int): the loss computed by the loss criterion.
```
Comment (francoishernandez) [inline onmt/utils/loss.py:246]: int?

Comment (vince62s) [reply onmt/utils/loss.py:246]: yes .sum().item() is done before passing now to avoid .clone()
